File: newsmodule/src/main/java/zjm/cst/dhu/newsmodule/mvp/presenter/NewsListPresenter.java
package zjm.cst.dhu.newsmodule.mvp.presenter;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import retrofit2.Retrofit;
import rx.Observable;
import rx.Observer;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.functions.Func2;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;
import zjm.cst.dhu.library.Constant;
import zjm.cst.dhu.library.api.repository.Repository;
import zjm.cst.dhu.library.event.FloatingActionButtonClickEvent;
import zjm.cst.dhu.library.mvp.model.NewsData;
import zjm.cst.dhu.library.utils.RetrofitUtil;
import zjm.cst.dhu.library.utils.RxBus;
import zjm.cst.dhu.newsmodule.mvp.view.NewsListContract;

/**
 * Created by zjm on 2017/5/10.
 */

public class NewsListPresenter implements NewsListContract.Presenter {
    private NewsListContract.View myView;
    private CompositeSubscription myCompositeSubscription;
    private int nowPosition = 0;
    private String mNewsId;
    private String mNewsType;

    @Inject
    public NewsListPresenter() {

    }

    @Override
    public void attachView(NewsListContract.View BaseView) {
        myView = BaseView;
        myCompositeSubscription = new CompositeSubscription();
    }

    @Override
    public void detachView() {
        if (myCompositeSubscription != null && myCompositeSubscription.isUnsubscribed()) {
            myCompositeSubscription.unsubscribe();
        }
    }

    @Override
    public void setValue(String id, String type) {
        this.mNewsId = id;
        this.mNewsType = type;
        subscribeFloatingActionButtonClickEvent();
    }

    private void subscribeFloatingActionButtonClickEvent(){
        Subscription subscription= RxBus.getDefault().toObservable(FloatingActionButtonClickEvent.class)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action1<FloatingActionButtonClickEvent>() {
                    @Override
                    public void call(FloatingActionButtonClickEvent floatingActionButtonClickEvent) {
                        myView.floatingActionButtonClick();
                    }
                });
        myCompositeSubscription.add(subscription);
    }

    @Override
    public void loadNewList() {
        nowPosition = 0;
        Retrofit retrofit = RetrofitUtil.setupRetrofit(RetrofitUtil.setupOkHttpClient(), Constant.NETEASE_NEWS_BASE_URL);
        Repository repository = new Repository(retrofit);
        Subscription subscription = repository.getNewsList(mNewsType, mNewsId, nowPosition)
                .flatMap(new Func1<Map<String, List<NewsData>>, Observable<NewsData>>() {
                    @Override
                    public Observable<NewsData> call(Map<String, List<NewsData>> stringListMap) {
                        if (mNewsId.endsWith("5rex5Zyz")) {
                            return Observable.from(stringListMap.get("深圳"));
                        }
                        return Observable.from(stringListMap.get(mNewsId));
                    }
                })
                .distinct()
                .toSortedList(new Func2<NewsData, NewsData, Integer>() {
                    @Override
                    public Integer call(NewsData newsData, NewsData newsData2) {
                        if (newsData.getAds() != null) {
                            return -1;
                        } else if (newsData2.getAds() != null) {
                            return 1;
                        }
                        return newsData2.getPtime().compareTo(newsData.getPtime());
                    }
                })
                .compose(new Observable.Transformer<List<NewsData>, List<NewsData>>() {
                    @Override
                    public Observable<List<NewsData>> call(Observable<List<NewsData>> listObservable) {
                        return listObservable
                                .unsubscribeOn(Schedulers.io())
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread());
                    }
                })
                .subscribe(new Observer<List<NewsData>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                        myView.loadNewListError();
                    }

                    @Override
                    public void onNext(List<NewsData> newsDataList) {
                        myView.loadNewListSuccess(newsDataList);
                        nowPosition+=Constant.NEWS_LIST_REFRESH_ITEM_MEDIUM_COUNTS;
                    }
                });
        myCompositeSubscription.add(subscription);
    }

    @Override
    public void loadNextList() {
        Retrofit retrofit = RetrofitUtil.setupRetrofit(RetrofitUtil.setupOkHttpClient(), Constant.NETEASE_NEWS_BASE_URL);
        Repository repository = new Repository(retrofit);
        Subscription subscription = repository.getNewsList(mNewsType, mNewsId, nowPosition)
                .flatMap(new Func1<Map<String, List<NewsData>>, Observable<NewsData>>() {
                    @Override
                    public Observable<NewsData> call(Map<String, List<NewsData>> stringListMap) {
                        if (mNewsId.endsWith("5rex5Zyz")) {
                            return Observable.from(stringListMap.get("深圳"));
                        }
                        return Observable.from(stringListMap.get(mNewsId));
                    }
                })
                .distinct()
                .toSortedList(new Func2<NewsData, NewsData, Integer>() {
                    @Override
                    public Integer call(NewsData newsData, NewsData newsData2) {
                        if (newsData.getAds() != null) {
                            return -1;
                        } else if (newsData2.getAds() != null) {
                            return 1;
                        }
                        return newsData2.getPtime().compareTo(newsData.getPtime());
                    }
                })
                .compose(new Observable.Transformer<List<NewsData>, List<NewsData>>() {
                    @Override
                    public Observable<List<NewsData>> call(Observable<List<NewsData>> listObservable) {
                        return listObservable
                                .unsubscribeOn(Schedulers.io())
                                .subscribeOn(Schedulers.io())
                                .observeOn(AndroidSchedulers.mainThread());
                    }
                })
                .subscribe(new Observer<List<NewsData>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                        myView.loadNextListError();
                    }

                    @Override
                    public void onNext(List<NewsData> newsDataList) {
                        myView.loadNextListSuccess(newsDataList);
                        nowPosition+=Constant.NEWS_LIST_REFRESH_ITEM_MEDIUM_COUNTS;
                    }
                });
        myCompositeSubscription.add(subscription);
    }
}

File: newsmodule/src/main/kotlin/zjm.cst.dhu.newsmodule/mvp/presenter/NewsListPresenter.kt
package zjm.cst.dhu.newsmodule.mvp.presenter

import javax.inject.Inject

import retrofit2.Retrofit
import rx.Observable
import rx.Observer
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action1
import rx.functions.Func1
import rx.functions.Func2
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import zjm.cst.dhu.library.Constant
import zjm.cst.dhu.library.api.repository.Repository
import zjm.cst.dhu.library.event.FloatingActionButtonClickEvent
import zjm.cst.dhu.library.mvp.model.NewsData
import zjm.cst.dhu.library.utils.network.RetrofitUtil
import zjm.cst.dhu.library.utils.others.RxBus
import zjm.cst.dhu.newsmodule.mvp.view.NewsListContract

/**
 * Created by zjm on 2017/5/10.
 */

class NewsListPresenter @Inject constructor()
    : NewsListContract.Presenter {
    private var myView: NewsListContract.View? = null
    private var myCompositeSubscription: CompositeSubscription? = null
    private var nowPosition = 0
    private var mNewsId: String? = null
    private var mNewsType: String? = null

    override fun attachView(BaseView: NewsListContract.View) {
        myView = BaseView
        myCompositeSubscription = CompositeSubscription()
    }

    override fun detachView() {
        if (myCompositeSubscription != null && myCompositeSubscription!!.isUnsubscribed) {
            myCompositeSubscription!!.unsubscribe()
        }
    }

    override fun setValue(id: String, type: String) {
        this.mNewsId = id
        this.mNewsType = type
        subscribeFloatingActionButtonClickEvent()
    }

    private fun subscribeFloatingActionButtonClickEvent() {
        val subscription = RxBus.default.toObservable(FloatingActionButtonClickEvent::class.java)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    myView!!.floatingActionButtonClick()
                })
        myCompositeSubscription!!.add(subscription)
    }

    override fun loadNewList() {
        nowPosition = 0
        if (mNewsId != null && mNewsType != null) {
            myCompositeSubscription!!.add(
                    prepareLoadList()
                            .subscribe(object : Observer<List<NewsData>> {
                                override fun onCompleted() {

                                }

                                override fun onError(e: Throwable) {
                                    e.printStackTrace()
                                    myView!!.loadNewListError()
                                }

                                override fun onNext(newsDataList: List<NewsData>) {
                                    myView!!.loadNewListSuccess(newsDataList)
                                    nowPosition += Constant.NEWS_LIST_REFRESH_ITEM_MEDIUM_COUNTS
                                }
                            })
            )
        }
    }

    override fun loadNextList() {
        if (mNewsId != null && mNewsType != null) {
            myCompositeSubscription!!.add(
                    prepareLoadList()
                            .subscribe(object : Observer<List<NewsData>> {
                                override fun onCompleted() {

                                }

                                override fun onError(e: Throwable) {
                                    e.printStackTrace()
                                    myView!!.loadNextListError()
                                }

                                override fun onNext(newsDataList: List<NewsData>) {
                                    myView!!.loadNextListSuccess(newsDataList)
                                    nowPosition += Constant.NEWS_LIST_REFRESH_ITEM_MEDIUM_COUNTS
                                }
                            })
            )
        }
    }

    private fun prepareLoadList():
            Observable<List<NewsData>> {
        val retrofit = RetrofitUtil.setupRetrofit(RetrofitUtil.setupOkHttpClient(),
                Constant.NETEASE_NEWS_BASE_URL)
        val repository = Repository(retrofit)
        val subscription = repository.getNewsList(mNewsType!!, mNewsId!!, nowPosition)
                .flatMap(Func1<Map<String, List<NewsData>>, Observable<NewsData>> { stringListMap ->
                    if (mNewsId!!.endsWith("5rex5Zyz")) {
                        return@Func1 Observable.from<NewsData>(stringListMap["深圳"])
                    }
                    Observable.from<NewsData>(stringListMap[mNewsId!!])
                })
                .distinct()
                .toSortedList(Func2<NewsData, NewsData, Int> { newsData, newsData2 ->
                    if (newsData.ads != null) {
                        return@Func2 -1
                    } else if (newsData2.ads != null) {
                        return@Func2 1
                    }
                    val t2 = newsData2.ptime
                    val t1 = newsData.ptime
                    if (t1 != null && t2 != null) {
                        t2.compareTo(t1)
                    } else {
                        0
                    }
                })
                .compose({ listObservable ->
                    listObservable
                            .unsubscribeOn(Schedulers.io())
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                })
        return subscription
    }
}
